src/files/filesProcessor.ts

```diff
@@ -0,0 +1,162 @@
+import { EgoJwtManager } from "auth";
+import { getFilesByProgramId } from "files/getFilesByProgramId";
+import _ from "lodash";
+import logger from "logger";
+import { initializeRdpcInfo } from "rdpc/analysesProcessor";
+import { DonorInfoMap, StringMap } from "rdpc/types";
+import {
+  DonorMolecularDataReleaseStatus,
+  File,
+  FileReleaseState,
+} from "./types";
+
+/** Gets file data from file-service, group files by donorId, and determine
+ *  donor's file release status based on the number of file status.
+ */
+export const determineReleaseStatus = async (
+  programId: string,
+  egoJwtManager: EgoJwtManager,
+  filesFetcher: typeof getFilesByProgramId,
+  donorIds?: string[]
+): Promise<DonorInfoMap> => {
+  const result: DonorInfoMap = {};
+  // gets files grouped by donor id
+  const files = await getFilesByPage({
+    programId: programId,
+    egoJwtManager: egoJwtManager,
+    filesFetcher: filesFetcher,
+    donorIds: donorIds,
+  });
+
+  Object.entries(files).forEach(([donorId, files]) => {
+    const filesByReleaseState = _.groupBy(files, (file) => file.releaseState);
+    const releaseStates = Object.keys(filesByReleaseState);
+
+    initializeRdpcInfo(result, donorId);
+
+    switch (releaseStates.length) {
+      case 1:
+        // FULLY_RELEASED - all files for donor have release_state of PUBLIC
+        if (releaseStates[0] === FileReleaseState.PUBLIC) {
+          result[donorId].releaseStatus =
+            DonorMolecularDataReleaseStatus.FULLY_RELEASED;
+        }
+
+        // NO_RELEASE - all files for donor have release_state of RESTRICTED or QUEUED
+        if (
+          releaseStates[0] === FileReleaseState.QUEUED ||
+          releaseStates[0] === FileReleaseState.RESTRICTED
+        ) {
+          result[donorId].releaseStatus =
+            DonorMolecularDataReleaseStatus.NO_RELEASE;
+        }
+        break;
+
+      case 2:
+        // If any file is in PUBLIC state, the donor file release state will be PARTIALLY_RELEASED.
+        if (releaseStates.indexOf(FileReleaseState.PUBLIC) != -1) {
+          result[donorId].releaseStatus =
+            DonorMolecularDataReleaseStatus.PARTIALLY_RELEASED;
+        } else {
+          result[donorId].releaseStatus =
+            DonorMolecularDataReleaseStatus.NO_RELEASE;
+        }
+        break;
+
+      case 3:
+        // PARTIALLY_RELEASED - a mix of PUBLIC and RESTRICTED or QUEUED amongst their files.
+        result[donorId].releaseStatus =
+          DonorMolecularDataReleaseStatus.PARTIALLY_RELEASED;
+        break;
+
+      default:
```

Comment:
its strange to have an inaccessible code block. either lets add a `logger.error` statement to the default block that can log out something like "more release statuses in donor's files than should be possible", or you can handle the 3-or-more case by doing:

```js
case 3:
// Intentionally fall through to default so we can handle 3 or more states. More than 3 states should never occur.
default:
        // PARTIALLY_RELEASED - a mix of PUBLIC and RESTRICTED or QUEUED amongst their files.
        result[donorId].releaseStatus =
          DonorMolecularDataReleaseStatus.PARTIALLY_RELEASED;
        break;
```